From 62af4b71196f06ebcccced00f9bbfdbc1544ab00 Mon Sep 17 00:00:00 2001
From: Finley Xiao <finley.xiao@rock-chips.com>
Date: Fri, 8 Oct 2021 14:32:39 +0800
Subject: [PATCH 07/19] rockchip: rk3308: Disable cpu idle for rk3308bs

The system will freeze or panic when stress tests if enable cpu idle.

Change-Id: I28e572afaec3a0c284d3d7df00e90921978652a5
Signed-off-by: Finley Xiao <finley.xiao@rock-chips.com>
---
 arch/arm/mach-rockchip/rk3308/rk3308.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/arch/arm/mach-rockchip/rk3308/rk3308.c b/arch/arm/mach-rockchip/rk3308/rk3308.c
index ab62457..f2bf29a 100644
--- a/arch/arm/mach-rockchip/rk3308/rk3308.c
+++ b/arch/arm/mach-rockchip/rk3308/rk3308.c
@@ -204,6 +204,29 @@ int arch_cpu_init(void)
 }
 #endif
 
+static int fdt_fixup_cpu_idle(const void *blob)
+{
+	int cpu_node, sub_node, len;
+	u32 *pp;
+
+	cpu_node = fdt_path_offset(blob, "/cpus");
+	if (cpu_node < 0) {
+		printf("Failed to get cpus node\n");
+		return -EINVAL;
+	}
+
+	fdt_for_each_subnode(sub_node, blob, cpu_node) {
+		pp = (u32 *)fdt_getprop(blob, sub_node, "cpu-idle-states",
+					&len);
+		if (!pp)
+			continue;
+		if (fdt_delprop((void *)blob, sub_node, "cpu-idle-states") < 0)
+			printf("Failed to del cpu-idle-states prop\n");
+	}
+
+	return 0;
+}
+
 static int fdt_fixup_cpu_opp_table(const void *blob)
 {
 	int opp_node, old_opp_node;
@@ -368,6 +391,7 @@ static int fdt_fixup_thermal_zones(const void *blob)
 int rk_board_fdt_fixup(const void *blob)
 {
 	if (soc_is_rk3308bs()) {
+		fdt_fixup_cpu_idle(blob);
 		fdt_fixup_cpu_opp_table(blob);
 		fdt_fixup_dmc_opp_table(blob);
 		fdt_fixup_thermal_zones(blob);
-- 
2.7.4

